### PR TITLE
Implementation of fmcw_range

### PR DESCRIPTION
**Pull request by enchom**

### CODE REVIEW

File: computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/ComplexVector.java [hunk 1]
```diff
@@ -0,0 +1,382 @@
+package uk.ac.cam.cl.juliet.computationengine.range;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.math3.complex.Complex;
+
+// TODO : Unit testing of everything
```
Comment (patricoferris) [inline computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/ComplexVector.java:7]: A shared todo task 😆

File: computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/WindowFunction.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package uk.ac.cam.cl.juliet.computationengine.range;
+
+import java.util.ArrayList;
+
+/** An interface defining a Window Function */
+public interface WindowFunction {
```
Comment (patricoferris) [inline computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/WindowFunction.java]: Don't know what your thoughts are on naming interfaces `IClassName` or not? I did it for one of mine and I guess we can either go with it or without it?

Comment (RosieBaish) [reply computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/WindowFunction.java]: I think we should

File: computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/BlackmanWindow.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    @Override
+    public ArrayList<Double> generateWindow(int n) {
+        ArrayList<Double> window = new ArrayList<>();
+
+        int m = (n + 1) / 2;
+        for (int i = 0; i < m; i++) {
+            window.add(evaluateFunction(n, i));
+        }
+
+        for (int i = m - 1 - (n % 2); i >= 0; i--) {
```
Comment (RosieBaish) [inline computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/BlackmanWindow.java:39]: Can you either add a comment to explain what's going on here, or switch it around.
Other option is to have the loop from m to n and then use logic to calculate the correct value.

File: computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/ComplexVector.java [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+        return addConstant(new Complex(c));
+    }
+
+    /**
+     * Creates a new vector that is the equal to {@code vector2} subtracted of {@code this}.
+     *
+     * @param vector2 A {@link ComplexVector} vector to be subtracted by {@code this}
+     * @return The new resulting {@link ComplexVector}
+     */
+    public ComplexVector subtractElements(ComplexVector vector2) {
```
Comment (RosieBaish) [inline computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/ComplexVector.java:156]: Do we actually need both add and subtract?

**Review by RosieBaish — APPROVED** [review]

File: computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/WindowFunction.java (hunk 2 above)
Comment (enchom) [reply computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/WindowFunction.java]: Agreed. Will change to `IWindowFunction`.

File: computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/BlackmanWindow.java (hunk 3 above)
Comment (enchom) [reply computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/BlackmanWindow.java:39]: The latter does seem more logical from code-readability point-of-view. Will change it.

File: computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/ComplexVector.java (hunk 4 above)
Comment (enchom) [reply computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/ComplexVector.java:156]: We technically don't, as add(negate(X)) == subtract(X), however I tried to make the class as general as possible and easy-to-use. Some of the functions are not even currently used but can be useful in the future. I think having a subtract function doesn't hurt - it's short, slightly more efficient than a combination of add+negate, and more natural to use.

Comment (RosieBaish) [reply computationEngine/src/main/java/uk/ac/cam/cl/juliet/computationengine/range/ComplexVector.java:156]: Makes sense. I agree.

**Pull request merged**